### PR TITLE
feat(tekton): Define default pipelines, override steps and agents in defaults

### LINKED ISSUE #3961: Add the ability to remove a whole pipeline or lifecycle
### Summary 

In some cases we need to remove a whole pipeline or a lifecycle inherited from a buildpack. Right now, the only way to do that is either by creating a new pipeline from scratch or by creating a bogus replacement (e.g., `echo nothing`).

### Jx version

Any

### Jenkins type

- [ ] Next Generation (Tekton + Prow)
- [ ] Classic Jenkins
- [X] Serverless Jenkins (JenkinsFileRunner + Prow)

### Kubernetes cluster

Any

### Operating system / Environment

Any

### PR DESCRIPTION
**Pull request by abayer**
#### Submitter checklist

- [x] Change is code complete and matches issue description.
- [x] Change is covered by existing or new tests.

#### Description

This adds a new field to `pipelineConfig.pipelines`: `default`. That
contains a `ParsedPipeline`, like
`pipelineConfig.pipelines.release.pipeline`, and is used as the
default for all pipelines (i.e., `release`, `pullRequest`, `feature`)
in the project/build pack if one isn't specified there already. It can
be overridden/added to via `pipelineConfig.pipelines.overrides`, which
also now has the `type` field, taking `replace`, `before`, and
`after`, defaulting to `replace` - if it's `replace`, the target step
is replaced with the new step or steps, while if it's `before` or
`after`, the new step or steps are put either before or after the
target step, as appropriate.

Example syntax in `jenkins-x.yml` or a build pack's `pipeline.yaml`
for defining a default pipeline:

```
pipelineConfig:
  pipelines:
    default:
      agent:
        image: go
      stages:
        - name: First Stage
          steps:
            - command: echo
              args: ['hello', 'there']
              name: hello
        - name: Second Stage
          steps:
            - command: echo
              args: ['goodbye']
              name: goodbye
```

Example override syntax (works with build packs with pipelines defined
as `default`, as `pipelineConfig.pipelines.release.pipeline`, or
standard lifecycles like `pipelineConfig.pipelines.release.build` etc):

```
pipelineConfig:
  pipelines:
    overrides:
      - pipeline: release
        stage: First Stage
        name: hello
        step:
          command: echo before we say hello...
        type: before
```

Also allows removing whole stages/lifecycles and pipelines,
and replacing/prepending/appending replacement steps for a whole stage.

If `name` is not given for an `override`, the whole stage (if `stages`
is specified) or whole pipeline (if just `pipelines` is specified) is
affected - either removed if no `step` or `steps` is given, or, for
stages, the existing steps are replaced/prepended/appended depending on `type`.

In addition, allow overriding agent on stage or pipeline from default Tekton pipelines. Specify an `agent` to override the existing `agent` for either a `stage` or a `pipeline`. Can be specified along with `step` or `steps`, or on its own.

```
pipelineConfig:
  pipelines:
    overrides:
      - pipeline: release
        stage: Some Stage
        agent:
          image: some-image
```

#### Special notes for the reviewer(s)

/assign @jstrachan 
/assign @rawlingsj 
/assign @warrenbailey 
/assign @dwnusbaum 
/assign @pmuir 
cc @joseblas @vfarcic @sharepointoscar 

#### Which issue this PR fixes

fixes #3961

### CODE REVIEW

**Comment by abayer** [comment]
/hold

Putting this on hold for the moment because it's a big enough pile of changes that I want to be sure there's consensus before it gets merged.

**Comment by abayer** [comment]
/assign @hferentschik 

Adding Hardy to make sure my mild hackery of codegen to generate the `DeepCopy` functions for `ProjectConfig`, `PipelineConfig`, `ParsedPipeline`, and friends isn't completely insane.

**Comment by abayer** [comment]
Oops, note for myself - overriding a default Pipeline defined in the jenkins-x.yml isn’t actually implemented. Will do that tomorrow am.

File: Makefile.codegen [hunk 1]
```diff
@@ -45,6 +45,7 @@ generate-client: codegen-clientset fmt ## Generate the client
 codegen-clientset: build-codegen ## Generate the k8s types and clients
 	@echo "Generating Kubernetes Clients for pkg/apis in pkg/client for jenkins.io:v1"
 	./build/$(CODE_GEN_BIN_NAME) --generator-version $(CLIENTSET_GENERATOR_VERSION) clientset --output-package=pkg/client --input-package=pkg/apis --group-with-version=jenkins.io:v1
+	./build/$(CODE_GEN_BIN_NAME) --generator-version $(CLIENTSET_GENERATOR_VERSION) clientset --generator deepcopy --output-package=pkg --input-package=pkg --group-with-version=config: --group-with-version=jenkinsfile: --group-with-version=tekton:syntax
```
Comment (hferentschik) [inline Makefile.codegen]: 👍

**Comment by hferentschik** [comment]
> Adding Hardy to make sure my mild hackery of codegen to generate the DeepCopy functions for ProjectConfig, PipelineConfig, ParsedPipeline, and friends isn't completely insane.

Looking good to me. Good to see that it was relatively easy to add the deepcopy.

**Comment by abayer** [comment]
@hferentschik - of interest to you, i've just added `--effective-pipeline` to the options for `jx step create task`, which will log out the "effective" `ParsedPipeline` before CRD generation but after build pack application, overrides, etc. Feels like it could be handy. =)

**Comment by warrenbailey** [comment]
This is awesome and will make our build packs much simpler and easier to use. 💯 

I just want to check my understanding:

1. A default pipeline is used in all pipelines, e.g. pull request, release etc, and can be defined in either a build pack or the jenkins-x.yml of the project
1. Overrides are defined in the jenkins-x.yml of the project and can be used to replace the default pipelines or append/prepend to it.
1. If a build pipeline exists in a jenkins-x.yml then that completely overrides the default pipeline.

**Comment by abayer** [comment]
@warrenbailey - 

1. A default pipeline is used for `release`/`pullRequest`/`feature` if and only if `release`/`pullRequest`/`feature` is not explicitly defined, and can be defined in either a build pack or `jenkins-x.yml`.
2. `overrides` can be defined in either `jenkins-x.yml` or a build pack - so a build pack that extends another build pack can override steps from the parent build pack, a `jenkins-x.yml` can override steps from its build pack, or a `jenkins-x.yml` can override steps from its `default` pipeline. And yes, "overriding" steps means any of replacing named steps, appending after named steps, or prepending before named steps.
3. Ok, actually addressed this in 1. =)

I'm not in love with the naming of `default` or `overrides` - `overrides` is already live, but it's early enough that we can still rename it.

**Comment by abayer** [comment]
This now addresses #3961 as well - see the description for details.

**Comment by abayer** [comment]
Oh, and `sh: ...` will now be legal in `jenkins-x.yml`/`default` pipelines as well as in build packs, because...well, I got annoyed and regretted making it not legal there. =)

**Comment by abayer** [comment]
@warrenbailey @vfarcic @sharepointoscar - so terminology - do any of y'all have suggestions to replace `default` and `overrides`? As I said, I don't love the wording and I know others don't either, but I can't think of anything better.

**Comment by vfarcic** [comment]
`default` looks OK to me.

I'm not sure why we need `overrides`. People are already used to `replace: true`. Shouldn't we continue using that logic?

I'm worried about us introducing too much complexity into pipelines. We can easily end up with YAML-pipelines not being easier to create than tekton-definitions. If that happens there will be no need for jx. I believe that we need to favor simplicity over anything else.

**Comment by abayer** [comment]
@vfarcic the buildpack-style syntax with `replace: true` only really fits for the older build pack with predefined lifecycles, or at least I couldn’t figure out a way to make it work for replacing individual steps, removing steps or stages/lifecycles, adding steps within a stage other than at the very beginning of the stage, or working with arbitrary stages in general. That said, I do think it’s possible I overshot in terms of the complexity - let me think about whether I can rework `overrides` to fit more directly into the `pipeline` syntax.

And fwiw, while we are definitely adding some complexity no matter what (since any new capabilities in the syntax inherently mean more complexity) and we certainly should err on the side of simplicity, I don’t think we’re in danger of reaching a point where our syntax is no easier to write than Tekton. Things like `overrides` (in whatever form it ends up taking) are optional, while Tekton’s syntax is always complex. =)

**Comment by abayer** [comment]
So one thing I definitely think is overly complex in `overrides` is `pipelines` and `stages` being slices - that was a real piece of premature optimization on my part. I'm going to change those to be simple string fields.

**Comment by warrenbailey** [comment]
> @warrenbailey @vfarcic @sharepointoscar - so terminology - do any of y'all have suggestions to replace `default` and `overrides`? As I said, I don't love the wording and I know others don't either, but I can't think of anything better.

I can't think of anything better, so let's go with it 👍

**Comment by abayer** [comment]
/hold cancel

Alright, I'm calling this "close enough" and taking it off hold.

**Comment by warrenbailey** [comment]
Cool all looks good, it'll be awesome when this is in :)

**Comment by vfarcic** [comment]
This was merged 5 hours ago and the last release is from 2 days. Is the process that creates releases down?

**Comment by vfarcic** [comment]
@abayer  Is this change available in the release `2.0.145`? If it is, I made the following pipeline:

```yaml
buildPack: go
pipelineConfig:
  pipelines:
    overrides:
      - pipeline: pullRequest
        type: replace
```

The pipeline does not even start. Did I make a mistake in syntax or there is an issue?

The following appears in pipelinerunner logs.

```bash
kubectl logs -l app=pipelinerunner
```

```
...
{"level":"error","msg":"failed to load project config in dir /tmp/git340032208: Validation failures in YAML file /tmp/git340032208/jenkins-x.yml:\npipelineConfig.pipelines.overrides.0: Additional property pipeline is not allowed\npipelineConfig.pipelines.overrides.0: Additional property type is not allowed\npipelineConfig.pipelines.overrides.1: Additional property pipeline is not allowed\npipelineConfig.pipelines.overrides.1: Additional property stage is not allowed\npipelineConfig.pipelines.overrides.1: Additional property type is not allowed failed to load project config in dir /tmp/git340032208: Validation failures in YAML file /tmp/git340032208/jenkins-x.yml:\npipelineConfig.pipelines.overrides.0: Additional property pipeline is not allowed\npipelineConfig.pipelines.overrides.0: Additional property type is not allowed\npipelineConfig.pipelines.overrides.1: Additional property pipeline is not allowed\npipelineConfig.pipelines.overrides.1: Additional property stage isnot allowed\npipelineConfig.pipelines.overrides.1: Additional property type is not allowed","time":"2019-05-20T21:40:16Z"}
```

My best guess is that pipelinerunner image is not up-to-date (I made a fresh cluster).

**Comment by abayer** [comment]
@vfarcic Yeah, looks like it hasn’t made it into pipelinerunner yet...

**Comment by abayer** [comment]
I think https://github.com/jenkins-x/jenkins-x-versions/pull/266 needs to get merged...

**Comment by sharepointoscar** [comment]
Looks like it got merged, safe to try this on a fresh cluster?

**Comment by vfarcic** [comment]
I created a following pipeline:

```yaml
buildPack: go
pipelineConfig:
  pipelines:
    overrides:
    - pipeline: pullRequest
      stage: build
      name: unit-tests
      step:
        command: make unittest
      type: before
```

However, when I create a PR, the output of the activities is:

```
STEP                                                 STARTED AGO DURATION STATUS
vfarcic/go-demo-6/PR-101 #1                                 3m5s      59s Succeeded
  from build pack                                           3m5s      59s Succeeded
    Credential Initializer Jrbjl                            3m5s       0s Succeeded
    Working Dir Initializer Dgdwk                           3m5s       0s Succeeded
    Place Tools                                             3m4s       1s Succeeded
    Build Container Build                                   3m2s      22s Succeeded
    Build Make Linux                                        3m2s      18s Succeeded
    Git Merge                                               3m2s       3s Succeeded
    Git Source Vfarcic Go Demo 6 Pr 101 Server Mxm9g        3m3s       1s Succeeded https://github.com/vfarcic/go-demo-6
    Postbuild Post Build                                    3m2s      23s Succeeded
    Promote Jx Preview                                      3m1s      55s Succeeded
    Promote Make Preview                                    3m1s      37s Succeeded
  Preview                                                   2m7s           https://github.com/vfarcic/go-demo-6/pull/101
    Preview Application                                     2m7s           http://go-demo-6.cd-vfarcic-go-demo-6-pr-101.34.74.132.38.nip.io
```

I don't see the new step (I checked the logs as well). I'm sure I made a mistake but I'm not sure where. Can you please let me know what I did wrong?

Just in case you need the version...

```
NAME               VERSION
jx                 2.0.253
jenkins x platform 2.0.330
Kubernetes cluster v1.12.8-gke.6
kubectl            v1.14.3
helm client        Client: v2.14.1+g5270352
git                git version 2.20.1 (Apple Git-117)
Operating System   Mac OS X 10.14.5 build 18F203
```

**Pull request merged**